Framework/DataObjects/src/Peak.cpp

```diff
@@ -168,14 +168,14 @@ Peak::Peak(const Geometry::IPeak &ipeak)
     : BasePeak(ipeak), m_detectorID(ipeak.getDetectorID()),
       m_initialEnergy(ipeak.getInitialEnergy()),
       m_finalEnergy(ipeak.getFinalEnergy()) {
-  setInstrument(ipeak.getInstrument());
+  const auto *peak = dynamic_cast<const Peak *>(&ipeak);
+  if (peak)
+    setInstrument(peak->getInstrument());
   detid_t id = ipeak.getDetectorID();
-  if (id >= 0) {
+  if (id >= 0)
     setDetectorID(id);
```

Comment:
`setDetectorID` will [throw](https://github.com/mantidproject/mantid/blob/a58b2490e272daa2d05d401287c4df5b23c5f082/Framework/DataObjects/src/Peak.cpp#L217) if the instrument is not set first which you have protected by `if (peak)`. I think it may be best to just have the constructor throw (with a helpful message) if it can't cast to `Peak` because there is no real way to make this work otherwise. We have an explicit constructor for making a `Peak` from `LeanElasticPeak`.